File: 2020/Day 2/Python/Solution.py
import re

with open('..\\input.txt') as file:
    lines = file.readlines()

def PartOne():
    wordsThatMatch = 0

    for a in lines:
        lowNum = int(a.split('-')[0])
        highNum = int((a.split('-')[1].split(' '))[0])
        charToFind = (((a.split(' '))[1]).split(':')[0])
        password = (a.split(": "))[1]

        charTotal = len(re.findall(charToFind, password))

        if charTotal >= lowNum and charTotal <= highNum:
            wordsThatMatch = wordsThatMatch + 1
        
    print("Found " + str(wordsThatMatch))

def PartTwo():
    wordsThatMatch = 0

    for a in lines:
        lowNum = int(a.split('-')[0])
        highNum = int((a.split('-')[1].split(' '))[0])
        charToFind = (((a.split(' '))[1]).split(':')[0])
        password = (a.split(": "))[1]
        found = 0

        positions = [x for x, v in enumerate(password) if v == charToFind]

        for p in positions:
            p = p + 1
            if p == lowNum or p == highNum:
                found = found + 1

        if found == 1:
            wordsThatMatch = wordsThatMatch + 1
        
    print("Found " + str(wordsThatMatch))

PartOne()
PartTwo()

File: 2020/Day 2/JavaScript/Solution.js
const fs = require('fs');

function PartOne(array) {
    wordsThatMatch = 0
    array.forEach(e => {
        charTotal = 0
        lowNum = e.split('-')[0]
        highNum = (e.split('-')[1].split(' '))[0]
        charToFind = (((e.split(' '))[1]).split(':')[0])
        password = (e.split(": "))[1]

        for (var i = 0; i < password.length; i++) {
            p = password[i]
            if (p.match(charToFind)){
                charTotal++;
            }
        }

        if (charTotal >= lowNum && charTotal <= highNum) { 
            wordsThatMatch++;
        }

    })
    console.log(`Words that match ${wordsThatMatch}`)
}

function PartTwo(array) { 
    wordsThatMatch = 0
    array.forEach(e => {
        charTotal = 0
        lowNum = e.split('-')[0]
        highNum = (e.split('-')[1].split(' '))[0]
        charToFind = (((e.split(' '))[1]).split(':')[0])
        password = (e.split(": "))[1]

        currentChar = 1
        for (var i = 0; i < password.length; i++) {
            p = password[i]
            if (p.match(charToFind) && (currentChar == lowNum || currentChar == highNum )){
                charTotal++
            }
            currentChar++
        }

        if (charTotal == 1) { 
            wordsThatMatch++;
        }

    })
    console.log(`Words that match ${wordsThatMatch}`)
}

fs.readFile('..\\Input.txt', 'UTF-8', (err, data) => {
    if (err) {
        console.log(err)
        return
    }

    var array = data.split('\n')
    results = PartOne(array)
    results = PartTwo(array)
})